docs/src/components/quick-nav/QuickNav.tsx

```diff
@@ -0,0 +1,263 @@
+'use client';
+import * as React from 'react';
+import clsx from 'clsx';
+
+export function Root({ children, className, ...props }: React.ComponentProps<'div'>) {
+  const ref = React.useRef<HTMLDivElement>(null);
+
+  // When nav height is larger than viewport, we make document scroll push it around
+  // so it sticks to top or bottom depending on scroll direction as if it was a real
+  // physical object constrained by the viewport
+  React.useEffect(() => {
+    if (!ref.current) {
+      return undefined;
+    }
+
+    let top: number;
```

Comment:
It seems that this whole code block can be defined outside of the component, so it's not unnecessarily evaluated on every render.

Reply:
Hmm we are inside `useEffect` with empty deps, so it's not evaluated on every render anyway, right? I'd prefer to keep these close to the stateful stuff if it doesn't matter

Reply:
It's not run on every render but it is evaluated (new functions are being allocated, etc.)

Reply:
Is this right?
https://github.com/mui/base-ui/pull/804/commits/b087f7caa8e727a43a29ab095af61f484d8a9d4b?diff=split&w=1

Reply:
Fair, it's rendered once per page load so in my understanding it won't impact anything tangible